### PR TITLE
Escape # in RFC 5987 pattern

### PR DESCRIPTION
**Pull request by sandstrom**
### Summary

The # seems to be missing an escape. Pasting `/[^A-Za-z0-9!#$&+.^_`|~-]/` into `irb` should also reveal the issue (# isn't there).

```ruby

# current
/[^A-Za-z0-9!#$&+.^_`|~-]/.match('#') # => no match
/[^A-Za-z0-9!#$&+.^_`|~-]/.match('$') # => no match

# suggestion
/[^A-Za-z0-9!\#$&+.^_`|~-]/.match('#') # => match
/[^A-Za-z0-9!\#$&+.^_`|~-]/.match('$') # => match

```

### CODE REVIEW

**Review by composerinteralia — COMMENTED** [review]
Yeah, this fix looks right to me. It looks like `#$` is syntax for interpolating the value of a global variable into a regexp. I had no idea that was even a thing.

```rb
$hello = "hello"
/#$hello/
#=> /hello/
```

Could we add some test coverage around this as well?

**Review by georgeclaghorn — APPROVED** [review]
Good to go once we have a test covering escaped characters.

**Comment by sandstrom** [comment]
@georgeclaghorn I've added a test now. 

Some of the automated tests/checks fail, but they seem unrelated to my change.

Shall I squash everything together or do you use the squash-merge button?

File: actionpack/test/dispatch/content_disposition_test.rb
Comment (composerinteralia) [inline actionpack/test/dispatch/content_disposition_test.rb:32]: This is not strictly necessary, since `#$-` doesn't seem to have the same interpolation problem as other patterns. But it may we worth defensively escaping it anyway in case anybody ever adds another character right after the dollar sign that does cause interpolation (e.g. `#$@`):

```suggestion
      disposition = Http::ContentDisposition.new(disposition: :inline, filename: "argh+!\#$-123_|&^`~.jpg")
```

**Review by composerinteralia — COMMENTED** [review]
This looks good to me. Yes, please do squash your commits into one.

File: actionpack/test/dispatch/content_disposition_test.rb
Comment (sandstrom) [reply actionpack/test/dispatch/content_disposition_test.rb]: I tried to make this change, but rubocop complained. I can't see why (see attached photo).

image

Comment (composerinteralia) [reply actionpack/test/dispatch/content_disposition_test.rb:32]: It doesn't like the single quotes:

```
 actionpack/test/dispatch/content_disposition_test.rb:32:82: C: Style/StringLiterals: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
      disposition = Http::ContentDisposition.new(disposition: :inline, filename: 'argh+!#$-123_|&^`~.jpg')
                                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^
```

Comment (sandstrom) [reply actionpack/test/dispatch/content_disposition_test.rb:32]: I've added a commit, rubocop job pending…

**Comment by sandstrom** [comment]
@composerinteralia Thanks! I've force-pushed a new commit where it passes rubocop.

Should be good to go now I guess (applied your other suggestion too).

Out of curiosity, what does "he/him" in your profile mean? I've seen that type of text on some peoples twitter accounts, but I don't understand what it is. I'm not an English native speaker so maybe it's something that I don't know about.

**Comment by composerinteralia** [comment]
Looks good!

I am [sharing the pronouns](https://lgbtqia.ucdavis.edu/guide-pronouns-allies) I prefer for someone to use when speaking or writing about me. Feel free to send me an email if you would like to learn more.

**Comment by sandstrom** [comment]
Alright, thanks for helping out on this. I really like Rails, it's an awesome project! 😄 

Thanks for the link, an interesting read! 

Sounds needlessly complicated though, reminds me of the Japanese emperor that insisted on special pronoun (to separate him from the peasants). I'll stick to the convenient he/she/it/… pronouns I learnt in school.

**Comment by composerinteralia** [comment]
> Sounds needlessly complicated though

It is needed, and in my opinion not so complicated (especially compared to software development 😄). I really am happy to chat more about this with you if you’re interested in learning, but let’s do it outside GitHub so folks don’t get too many extra notifications.

**Comment by sandstrom** [comment]
@composerinteralia I'm not that interested, sorry. 

Thanks for helping out with this fix + for your work on rails in general! 🥇

**Pull request merged**
